src/renderer.ts

```diff
@@ -107,67 +94,74 @@ export class Renderer {
    * @param backend Renderer.Backends.CANVAS or Renderer.Backends.SVG
    */
-  constructor(canvasId: string | HTMLCanvasElement | HTMLDivElement, backend: number) {
-    if (!canvasId) {
-      throw new RuntimeError('BadArgument', 'Invalid id for renderer.');
-    } else if (typeof canvasId === 'string') {
-      this.elementId = canvasId;
-      this.element = document.getElementById(canvasId) as HTMLCanvasElement | HTMLDivElement;
-    } else if ('getContext' in canvasId /* HTMLCanvasElement */) {
-      this.element = canvasId as HTMLCanvasElement;
+  constructor(context: RenderContext);
+  constructor(canvas: string | HTMLCanvasElement | HTMLDivElement, backend: number);
+  constructor(arg0: string | HTMLCanvasElement | HTMLDivElement | RenderContext, arg1?: number) {
+    if (isRenderContext(arg0)) {
+      // The user has provided what looks like a RenderContext, let's just use it.
+      // TODO(tommadams): RenderContext is an interface, can we introduce a context base class
+      // to make this check more robust?
+      this.ctx = arg0;
     } else {
-      // Assume it's a HTMLDivElement.
-      this.element = canvasId as HTMLDivElement;
-    }
+      if (arg1 === undefined) {
+        // The backend must be specified if the render context isn't directly provided.
+        throw new RuntimeError('InvalidArgument', 'Missing backend argument');
+      }
+      const backend: number = arg1;
 
-    // Verify backend and create context
-    this.backend = backend;
-    if (this.backend === Renderer.Backends.CANVAS) {
-      const canvasElement = this.element as HTMLCanvasElement;
-      if (!canvasElement.getContext) {
-        throw new RuntimeError('BadElement', `Can't get canvas context from element: ${canvasId}`);
+      let element: HTMLElement;
+      if (typeof arg0 == 'string') {
+        const maybeElement = document.getElementById(arg0);
+        if (maybeElement == null) {
+          throw new RuntimeError('BadElementId', `Can't find element with ID "${maybeElement}"`);
+        }
+        element = maybeElement;
       } else {
-        const context = canvasElement.getContext('2d');
-        if (context) {
-          this.ctx = new CanvasContext(context);
-        } else {
-          throw new RuntimeError('BadElement', `Can't get canvas context from element: ${canvasId}`);
+        element = arg0 as HTMLElement;
+      }
+
+      // Verify backend and create context
+      if (backend === Renderer.Backends.CANVAS) {
+        if (!(element instanceof window.HTMLCanvasElement)) {
+          throw new RuntimeError('BadElement', 'CANVAS context requires an HTMLCanvasElement');
         }
+        const context = element.getContext('2d');
+        if (!context) {
+          throw new RuntimeError('BadElement', "Can't get canvas context");
+        }
+        this.ctx = new CanvasContext(context);
+      } else if (backend === Renderer.Backends.SVG) {
+        if (!(element instanceof HTMLDivElement)) {
```

Comment:
How come we use window.HTMLCanvasElement above, but here we use HTMLDivElement?

Should we drop the window. prefix, for situations where window is not defined, but those types are somehow available in the globalThis scope?

Reply:
(I saw your commit comment: "window.HTMLCanvasElement so code works under node.js").

Does jsdom have `window.HTMLCanvasElement`, but not `HTMLCanvasElement`? How come `HTMLDivElement ` works?

Reply:
I played around with jsdom. We can assume that HTMLCanvasElement and HTMLDivElement are globally available IF the developer does this in node:

```
const dom = new JSDOM(`<!DOCTYPE html><body></body>`);
global.window = dom.window;
global.document = dom.window.document;
global.HTMLCanvasElement = dom.window.HTMLCanvasElement;
global.HTMLDivElement = dom.window.HTMLDivElement;
```

Then these will log true:
```
const d = document.createElement('div');
console.log(d instanceof HTMLDivElement);
const c = document.createElement('canvas');
console.log(c instanceof HTMLCanvasElement);
```

So I guess it's safest if we add the `window.` prefix to all our instanceof checks, e.g., `instanceof window.HTMLDivElement`.

Or perhaps we should expect the developer to make these two types globally available? We could document this in our examples and on the wiki.

Reply:
> (I saw your commit comment: "window.HTMLCanvasElement so code works under node.js").
> 
> Does jsdom / node-canvas have `window.HTMLCanvasElement`, but not `HTMLCanvasElement`? How come `HTMLDivElement ` works?

Thanks for the catch, I missed updating `HTMLDivElement`, it doesn't actually work :)

I only caught the `HTMLCanvasElement` because it failed when I was ran the visual regression test and that (of course) doesn't exercise any of the SVG code paths.